elastic/kibana
# [Cases] Migrate user actions connector ID

Background: #105677

Initial PR: https://github.com/elastic/kibana/pull/104221

This is a follow on PR for the work to migrate references to saved objects that are stored in fields outside of the `references` field for user actions. The connector ID can be stored in a few types of user actions:

Creation of a case user action
```
{
...
  connector: {
    id: <connector id to be migrated>,
    ...
  }
}
```

Updating the connector within a case
```
{
  id: <id to be migrated>
  ...
}
```

Pushing a case
```
{
  connector_id: <id to be migrated>
  ...
}
```

User actions holds a JSON encoded version of the above objects in the fields named `oldValue` and `newValue` to indicate what the original value was and what it is being changed to.

## Performance Concern
The easiest solution would be to have the backend decode the user actions `oldValue` and `newValue` fields and place the connector ID back into them and re-encode the fields before they are returned to the UI. This would have avoiding needing to change any code in the UI. We had some performance concerns with doing that approach for every user action when the case detail view is requested.

Instead the approach we took was to move the ids to `new_val_connector_id` and `old_val_connector_id`. This required a number of changes in the UI to look for that field instead of using the encoded values for the id.

## Notable Changes
- Backend
  - Migration for user actions saved objects to extract the connector id and move it to the `references` field
  - The json encoding work was moved down into the service layer
  - The service layer handles remove the connector id field and moving it to the `references` field
  - When returning the user actions the connector id is placed in the `new_val_connector_id` and `old_val_connector_id` fields to correspond with the `newValue` and `oldValue`
- Frontend
  - Refactored the places where we were accessing the connector id from the encoded value to leverage both the encoded value and the appropriate `*_val_connector_id`
  - Added more type checking for places where we called `parseString` on the encoded value

## Testing
I added unit tests for the transformation and migration code in the backend and frontend where I could. I also added integration tests that run the migration on older 7.13 data.

To test this PR:
- Create a case
- Add/remove/update the connector
- Push the case
- Add comments and other attachments
- Ensure that the user actions are all displaying the type of action taken by the user correctly
